core/server/master/src/main/java/alluxio/master/backup/AbstractBackupRole.java

```diff
@@ -0,0 +1,209 @@
+
+  /** Journal system. */
+  protected JournalSystem mJournalSystem;
+
+  /** Backup Manager. */
+  protected BackupManager mBackupManager;
+
+  /** User information for server. */
+  protected UserState mServerUserState;
+
+  /** Used to track backup status. */
+  protected BackupTracker mBackupTracker;
+
+  /**
+   * Creates a BackupMaster.
+   *
+   * @param masterContext master context
+   */
+  public AbstractBackupRole(CoreMasterContext masterContext) {
+    // Executor service will be used for driving the backup and connections.
+    mExecutorService =
+        Executors.newCachedThreadPool(ThreadFactoryUtils.build("backup-executor-%d", true));
+    // Task scheduler will be used for various backup related tasks.
+    mTaskScheduler =
+        Executors.newScheduledThreadPool(1, ThreadFactoryUtils.build("backup-task-%d", true));
+    // Store master context objects.
+    mUfsManager = masterContext.getUfsManager();
+    mJournalSystem = masterContext.getJournalSystem();
+    mBackupManager = masterContext.getBackupManager();
+    mServerUserState = masterContext.getUserState();
+    // Initialize messaging infra.
+    initializeCatalystContext();
+    // Read properties.
+    mCatalystRequestTimeout =
+        ServerConfiguration.getMs(PropertyKey.MASTER_BACKUP_TRANSPORT_TIMEOUT);
+    // Initialize backup tracker.
+    mBackupTracker = new BackupTracker();
+  }
+
+  /**
+   * Initializes the catalyst context with known message types.
+   */
+  private void initializeCatalystContext() {
+    Serializer messageSerializer = new Serializer().register(BackupRequestMessage.class)
+        .register(BackupHeartbeatMessage.class).register(BackupSuspendMessage.class);
+    mCatalystContext = new SingleThreadContext("backup-context-%d", messageSerializer);
+  }
+
+  /**
+   * Used to send message via connection and wait until response is received. Response is discarded
+   * since backup messages are one-way.
+   */
+  protected void sendMessageBlocking(Connection connection, Object message) throws IOException {
+    try {
+      mCatalystContext.execute(() -> {
+        connection.sendAndReceive(message);
+      }).get();
+    } catch (InterruptedException ie) {
+      throw new RuntimeException("Interrupted while sending message.");
+    } catch (ExecutionException ee) {
+      throw new IOException("Failed to send message", ee.getCause());
+    }
+  }
+
+  /**
+   * Takes a backup on local master. Note: Master state must have been suspended or locked before
+   * calling this.
+   *
+   * @param request the backup request
+   * @param entryCounter counter to receive written entry count
+   * @return URI of the backup
+   */
+  protected AlluxioURI takeBackup(BackupPRequest request, AtomicLong entryCounter)
+      throws IOException {
+    AlluxioURI backupUri;
+    // Acquire the UFS resource under which backup is being created.
+    try (CloseableResource<UnderFileSystem> ufsResource =
+        mUfsManager.getRoot().acquireUfsResource()) {
```

Comment:
Are backups only written to the root ufs?

Reply:
Nice catch, I think a fix for that regression was checked in after I started this.

Reply:
Oh, Looks like I still have that fix. In line 152 it overrides `ufs` resource with local factory if local backup is requested.

Reply:
Do we allow the case where it is a nonlocal, nonroot UFS?

Reply:
https://github.com/Alluxio/alluxio/issues/10066